go/adbc/driver/snowflake/record_reader.go

```diff
@@ -596,80 +582,103 @@ func newRecordReader(ctx context.Context, alloc memory.Allocator, ld gosnowflake
-				rdr, err := batch.GetStream(ctx)
-				if err != nil {
-					return err
-				}
-				defer rdr.Close()
+		rr, err := ipc.NewReader(r, ipc.WithAllocator(alloc))
+		if err != nil {
+			return nil, adbc.Error{
+				Msg:  err.Error(),
+				Code: adbc.StatusInvalidState,
+			}
+		}
+
+		var recTransform recordTransformer
+		rdr.schema, recTransform = getTransformer(rr.Schema(), ld, useHighPrecision)
 
-				rr, err := ipc.NewReader(rdr, ipc.WithAllocator(alloc))
+		group.Go(func() error {
+			defer rr.Release()
+			defer r.Close()
+			if len(batches) > 1 {
+				defer close(ch)
+			}
+
+			for rr.Next() && ctx.Err() == nil {
+				rec := rr.Record()
+				rec, err = recTransform(ctx, rec)
 				if err != nil {
 					return err
 				}
-				defer rr.Release()
+				ch <- rec
+			}
+			return rr.Err()
+		})
+
+		chs[0] = ch
+
+		lastChannelIndex := len(chs) - 1
+		go func() {
+			for i, b := range batches[1:] {
+				batch, batchIdx := b, i+1
+				chs[batchIdx] = make(chan arrow.Record, bufferSize)
+				group.Go(func() error {
+					// close channels (except the last) so that Next can move on to the next channel properly
+					if batchIdx != lastChannelIndex {
+						defer close(chs[batchIdx])
+					}
 
-				for rr.Next() && ctx.Err() == nil {
-					rec := rr.Record()
-					rec, err = recTransform(ctx, rec)
+					rdr, err := batch.GetStream(ctx)
 					if err != nil {
 						return err
 					}
-					chs[batchIdx] <- rec
-				}
+					defer rdr.Close()
 
-				return rr.Err()
-			})
-		}
+					rr, err := ipc.NewReader(rdr, ipc.WithAllocator(alloc))
+					if err != nil {
+						return err
+					}
+					defer rr.Release()
 
-		// place this here so that we always clean up, but they can't be in a
-		// separate goroutine. Otherwise we'll have a race condition between
-		// the call to wait and the calls to group.Go to kick off the jobs
-		// to perform the pre-fetching (GH-1283).
-		rdr.err = group.Wait()
-		// don't close the last channel until after the group is finished,
-		// so that Next() can only return after reader.err may have been set
-		close(chs[lastChannelIndex])
-	}()
+					for rr.Next() && ctx.Err() == nil {
+						rec := rr.Record()
+						rec, err = recTransform(ctx, rec)
+						if err != nil {
+							return err
+						}
+						chs[batchIdx] <- rec
+					}
+
+					return rr.Err()
+				})
+			}
+
+			// place this here so that we always clean up, but they can't be in a
+			// separate goroutine. Otherwise we'll have a race condition between
+			// the call to wait and the calls to group.Go to kick off the jobs
+			// to perform the pre-fetching (GH-1283).
+			rdr.err = group.Wait()
+			// don't close the last channel until after the group is finished,
+			// so that Next() can only return after reader.err may have been set
+			close(chs[lastChannelIndex])
+		}()
+	} else {
+		schema, err := rowTypesToArrowSchema(ctx, ld, useHighPrecision)
+		if err != nil {
+			return nil, err
+		}
+		rdr.schema, _ = getTransformer(schema, ld, useHighPrecision)
+	}
```

Comment:
Can we make this a bit more readable by flipping the condition and then just returning early in this case?

```go
if len(batches) == 0 {
    schema, err := rowTypesToArrowSchema(ctx, ld, useHighPrecision)
    if err != nil {
        return nil, err
    }
    rdr.schema, _ = getTransformer(schema, ld, useHighPrecision)
    return rdr, nil
}

// rest of code for channel management
```

Reply:
we could even skip all of the channel creation etc. for this case too most likely